### PR TITLE
implement mkdir and rmdir, add stat crate

### PR DESCRIPTION
**Pull request by sajattack**

### CODE REVIEW

File: include/sys/stat.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
+#ifndef _STAT_H
+#define _STAT_H
+
+#include <sys/types.h>
+
+struct stat {
```
Comment (jackpot51) [inline include/sys/stat.h]: The right way to do this is to have a include/bits/stat.h that defines the stat struct, so the rest of this file can be automatically generated

File: src/platform/src/redox/mod.rs [hunk 2]
```diff
@@ -118,6 +119,20 @@ pub fn link(path1: *const c_char, path2: *const c_char) -> c_int {
     e(unsafe { syscall::link(path1.as_ptr(), path2.as_ptr()) }) as c_int
 }
 
+pub fn mkdir(path: *const c_char, mode: mode_t) -> c_int {
+    let flags = O_CREAT | O_EXCL | O_CLOEXEC | O_DIRECTORY | mode as usize & 0o777;
+    let path = unsafe { c_str(path) };
+    let fd = e(syscall::open(path, flags)) as c_int;
```
Comment (jackpot51) [inline src/platform/src/redox/mod.rs]: You could do the match on syscall::open()

```
match syscall::open(path, flags) {
    Ok(fd) => {
        syscall::close(fd);
        0
    },
    Err(err) => {
        e(Err(err)) as c_int
    }
}
```

File: src/stat/build.rs [hunk 3]
```diff
@@ -0,0 +1,13 @@
+extern crate cbindgen;
+
+use std::{env, fs};
+
+fn main() {
+    /*
+     *let crate_dir = env::var("CARGO_MANIFEST_DIR").expect("CARGO_MANIFEST_DIR not set");
```
Comment (jackpot51) [inline src/stat/build.rs]: You should reenable this when you make the change mentioned above

File: src/ctype/src/lib.rs [hunk 4]
```diff
@@ -80,18 +80,10 @@ pub extern "C" fn toascii(c: c_int) -> c_int {
 
 #[no_mangle]
 pub extern "C" fn tolower(c: c_int) -> c_int {
-    if isupper(c) != 0 {
-        c + 0x20
-    } else {
-        c
-    }
+    if isupper(c) != 0 { c + 0x20 } else { c }
```
Comment (dlrobertson) [inline src/ctype/src/lib.rs]: Our rustfmt.toml shouldn't allow this.

Comment (sajattack) [reply src/ctype/src/lib.rs]: I ran cargo fmt instead of ./fmt.sh by accident

Comment (dlrobertson) [reply src/ctype/src/lib.rs]: Ah np

**Comment by jackpot51** [comment]
This was rebased and merged

**Pull request closed without merging**